File: code/java/DesignPatterns/src/main/java/proxy/ProxyLegalService.java
package proxy;

class ProxyLegalService implements LegalService {
    private RealLegalService realLegalService;

    @Override
    public void requestLegalAssistance() {
        if (realLegalService == null) {
            realLegalService = new RealLegalService();
        }

        // Add additional logic done by the proxy
        System.out.println("Proxy legal service is requesting legal assistance.");
        realLegalService.provideLegalAssistance();
    }
}

File: code/csharp/DesignPatterns/proxy/ProxyLegalService.cs
using System;

namespace Proxy
{
    public class ProxyLegalService : ILegalService
    {
        private RealLegalService _realLegalService;

        public void RequestLegalAssistance()
        {
            if (_realLegalService == null)
            {
                _realLegalService = new RealLegalService();
            }

            Console.WriteLine("Proxy legal service is requesting legal assistance.");
            _realLegalService.ProvideLegalAssistance();
        }
    }
}
